pyvista/plotting/charts.py

```diff
@@ -0,0 +1,827 @@
+import numpy as np
+from typing import Tuple
+
+import pyvista
+from pyvista import _vtk
+from .tools import parse_color
+
+try:
+    import matplotlib.figure
+    from matplotlib.backends.backend_agg import FigureCanvasAgg
+    _HAS_MPL = True
+except ModuleNotFoundError:
+    _HAS_MPL = False
+
+#region Some metaclass wrapping magic until a wrapped vtkPlotPie object can be added to a vtkChartPie
+class _vtkWrapperMeta(type):
+
+    def __call__(cls, *args, _wrap=None, **kwargs):
+        # if _wrap is None:
+        #     obj = cls.__new__(cls, *args, **kwargs)
+        # else:
+        #     obj = cls.__new__(_wrap.__class__, _wrap.__this__, *args, **kwargs)
+        obj = cls.__new__(cls, *args, **kwargs)
+        obj._wrapped = _wrap
+        obj.__init__(*args, **kwargs)
+        return obj
+
+
+class _vtkWrapper(object, metaclass=_vtkWrapperMeta):
```

Comment:
Thanks for your comments, I will try to take these things into account!
I had another question regarding the wrapping of vtk objects. From what I have seen in the pyvista codebase, the vtk type is typically subclassed and the wrapped instance is passed around to the vtk methods. This works fine and I tried to follow this approach throughout this PR. But what should I do when I have an existing vtk object (created internally by the vtk library within a function call) and want to wrap it to add certain functionality? Is this already done somewhere within the pyvista codebase?
I noticed you can create a wrapped version of native vtk objects in the following way:
```py
obj = _vtk.vtkPlotPie()
wrapped = _vtk.vtkPlotPie(obj.__this__)
assert obj.__this__ == wrapped.__this__
```
Unfortunately, this doesn't seem to work for subclassed types. E.g. the following does not properly wrap the original object
```py
class PiePlot(_vtk.vtkPlotPie):
   pass
obj = PiePlot()
wrapped = PiePlot(obj.__this__)
assert obj.__this__ == wrapped.__this__  # Not correctly wrapping obj
```
For now I came up with this wrapping code using some metaclass and getattribute magic, but not sure whether this is the best approach...

Reply:
I had the same problem with the [`Light`](https://github.com/pyvista/pyvista/blob/main/pyvista/plotting/lights.py#L30) class. What I ended up doing was to [define a `classmethod`](https://github.com/pyvista/pyvista/blob/main/pyvista/plotting/lights.py#L1088-L1119) that makes a copy of the input vtk object and throws it away. As far as I can tell this is the price we have to pay for not using composition.

Related PR was https://github.com/pyvista/pyvista/pull/1040 though I can't see too much discussion of this detail there.

Reply:
Ok, I see. But I'm afraid a copy won't be working in this case, as the containing chart (vtkCharPie instance) still has a reference to the original vtkPlotPie object and uses this original object to draw the plot to the screen. So any changes a user makes on that copied instance, won't be reflected on the screen.
So for now, I'll keep the wrapping code. I will create a PR for the vtk codebase to allow us to (re)set the default generated vtkPlotPie object, such that we don't have to do this wrapping (and can pass a new wrapped PiePlot instance to the containing chart). It's weird that they didn't implement this already, as the exact same situation arises for boxplots but there they added a public method to replace the contained plot object...

Reply:
Another approach is what we used for the [RenderWindowInteractor](https://github.com/pyvista/pyvista/blob/abe056bbc97667c722eb50329cae90ac76fc253c/pyvista/plotting/render_window_interactor.py) where we assign the vtk object as an internal attribute and then wrap that.  It's not perfect, but it works, doesn't require a copy, etc.

Reply:
That would be the composition alternative (as opposed to inheritance). I usually prefer that, and I think that's what I wanted to do for `Light` originally ;)